File: mavsim_matlab/chap10/mavsim_chap10.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% mavsim_matlab 
%     - Chapter 10 assignment for Beard & McLain, PUP, 2012
%     - Update history:  
%         3/12/2019 - RWB
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
clear all, path(pathdef)  % clear all variables and reset path 
run('../parameters/simulation_parameters')  % load SIM: simulation parameters
run('../parameters/aerosonde_parameters')  % load MAV: aircraft parameters
run('../parameters/control_parameters')  % load CTRL: control parameters
run('../parameters/sensor_parameters')  % load SENSOR: sensor parameters

% initialize the viewer
addpath('../chap10'); path_view = path_viewer();
addpath('../chap3'); data_view = data_viewer(); % use for debugging

% initialize the video writer
VIDEO = 0;  % 1 means write video, 0 means don't write video
if VIDEO==1, video=video_writer('chap10_video.avi', SIM.ts_video); end

% initialize elements of the architecture
addpath('../chap7');  mav = mav_dynamics(SIM.ts_simulation, MAV);
addpath('../chap4');  wind = wind_simulation(SIM.ts_simulation);
addpath('../chap6');  ctrl = autopilot(SIM.ts_simulation);
addpath('../chap8');  obsv = observer(SIM.ts_simulation);
addpath('../chap10'); path_follow = path_follower();

% path definition
addpath('../message_types'); path = msg_path();
%path.flag = 'line';
path.flag = 'orbit';
if isequal(path.flag, 'line')
    path.line_origin = [0.0; 0.0; -100.0];
    path.line_direction = [0.5; 1.0; 0.0];
    path.line_direction = path.line_direction / norm(path.line_direction);
else  % path.flag == 'orbit'
    path.orbit_center = [0.0; 0.0; -100.0];  
    path.orbit_radius = 200.0;  
    path.orbit_direction = 'CW';  
end

% initialize the simulation time
sim_time = SIM.start_time;

% main simulation loop
disp('Type CTRL-C to exit');
while sim_time < SIM.end_time
    
    %-------observer-------------
    mav.update_sensors(MAV, SENSOR); % get sensor measurement
    measurements = mav.sensors;
    estimated_state = obsv.update(measurements, MAV);

    %-------path follower-----
    autopilot_commands = path_follow.update(path, estimated_state);

    %-------controller-------------
    [delta, commanded_state] = ctrl.update(autopilot_commands, estimated_state);
    % use true states for debugging
    %[delta, commanded_state] = ctrl.update(autopilot_commands, mav.true_state);
    
    %-------physical system-------------
    current_wind = wind.update();
    mav.update(delta, current_wind, MAV);
    
    %-------update viewer-------------
    path_view.update(path, mav.true_state);
    % use for debugging
%     data_view.update(mav.true_state,...
%                      estimated_state,...
%                      commanded_state,...
%                      SIM.ts_simulation); 
    if VIDEO==1, video.update(sim_time);  end

    %-------increment time-------------
    sim_time = sim_time + SIM.ts_simulation;
end

if VIDEO==1, video.close(); end

File: mavsim_python/launch_files/chap10/mavsim_chap10.py
"""
mavsim_python
    - Chapter 10 assignment for Beard & McLain, PUP, 2012
    - Last Update:
        3/11/2019 - RWB
        2/27/2020 - RWB
        1/5/2023 - David L. Christiansen
        7/13/2023 - RWB
"""
import os, sys
# insert parent directory at beginning of python search path
from pathlib import Path
sys.path.insert(0,os.fspath(Path(__file__).parents[2]))
# use QuitListener for Linux or PC <- doesn't work on Mac
#from tools.quit_listener import QuitListener
import numpy as np
import parameters.simulation_parameters as SIM
from models.mav_dynamics_sensors import MavDynamics
from models.wind_simulation import WindSimulation
from controllers.autopilot import Autopilot
from estimators.observer_full import Observer
#from estimators.observer import Observer
from planners.path_follower import PathFollower
from viewers.view_manager import ViewManager
from message_types.msg_path import MsgPath
import time

# initialize elements of the architecture
wind = WindSimulation(SIM.ts_simulation)
mav = MavDynamics(SIM.ts_simulation)
autopilot = Autopilot(SIM.ts_simulation)
observer = Observer(SIM.ts_simulation)
path_follower = PathFollower()
viewers = ViewManager(path=True, 
                      data=False,
                      video=False, video_name='chap10.mp4')
#quitter = QuitListener()

# path definition
# path = MsgPath(
#     type='line', 
#     airspeed = 25,
#     line_origin = np.array([[0.0, 0.0, -100.0]]).T,
#     line_direction = np.array([[0.5, 1.0, 0.0]]).T,
#     )
path = MsgPath(
    type='orbit', 
    airspeed = 25,
    orbit_center = np.array([[0.0, 0.0, -100.0]]).T,
    orbit_radius = 200.0,
    orbit_direction = 'CCW', # 'CCW',
    )
# path = MsgPath(
#     type='helix', 
#     airspeed = 25,
#     orbit_center = np.array([[0.0, 0.0, -100.0]]).T,
#     orbit_radius = 200.0,
#     orbit_direction = 'CW', # 'CCW',
#     helix_climb_angle=np.radians(1),
#     helix_start_angle=np.radians(0),
#     )

# initialize the simulation time
sim_time = SIM.start_time
end_time = 300

# main simulation loop
print("Press 'Esc' to exit...")
while sim_time < end_time:
    # -------observer-------------
    measurements = mav.sensors()  # get sensor measurements
    estimated_state = observer.update(measurements)  # estimate states from measurements

    # -------path follower-------------
    autopilot_commands = path_follower.update(path, estimated_state)
    #autopilot_commands = path_follower.update(path, mav.true_state)  # for debugging

    # -------autopilot-------------
    delta, commanded_state = autopilot.update(autopilot_commands, estimated_state)

    # -------physical system-------------
    current_wind = wind.update()  # get the new wind vector
    mav.update(delta, current_wind)  # propagate the MAV dynamics

    # -------- update viewer -------------
    viewers.update(
        sim_time,
        true_state=mav.true_state,  # true states
        estimated_state=estimated_state,  # estimated states        
        commanded_state=commanded_state,  # commanded states
        delta=delta, # inputs to MAV
        path=path, # path
    )
        
    # -------Check to Quit the Loop-------
    # if quitter.check_quit():
    #     break

    # -------increment time-------------
    sim_time += SIM.ts_simulation
    #time.sleep(0.001)  # make the sim run slower

# close viewers
viewers.close(dataplot_name="ch10_data_plot")





